json/src/test/java/org/apache/JsonGenericSerializerTest.java

```diff
@@ -0,0 +1,103 @@
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache;
+
+import static net.javacrumbs.jsonunit.assertj.JsonAssertions.assertThatJson;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.time.ZonedDateTime;
+import java.util.Optional;
+import java.util.UUID;
+
+import org.apache.dto.BaseType;
+import org.apache.dto.FirstDomainObject;
+import org.apache.dto.SecondDomainObject;
+import org.apache.dto.TestModules;
+import org.apache.james.json.DTO;
+import org.apache.james.json.JsonGenericSerializer;
+import org.junit.jupiter.api.Test;
+
+class JsonGenericSerializerTest {
+    FirstDomainObject FIRST = new FirstDomainObject(Optional.of(1L), ZonedDateTime.parse("2016-04-03T02:01+07:00[Asia/Vientiane]"), "first payload");
+    SecondDomainObject SECOND = new SecondDomainObject(UUID.fromString("4a2c853f-7ffc-4ce3-9410-a47e85b3b741"), "second payload");
+
+    String FIRST_JSON = "{\"type\":\"first\",\"id\":1,\"time\":\"2016-04-03T02:01+07:00[Asia/Vientiane]\",\"payload\":\"first payload\"}";
+    String SECOND_JSON = "{\"type\":\"second\",\"id\":\"4a2c853f-7ffc-4ce3-9410-a47e85b3b741\",\"payload\":\"second payload\"}";
+
+    @Test
+    void shouldDeserializeKnownType() throws Exception {
+        assertThat(JsonGenericSerializer.of(TestModules.FIRST_TYPE)
+            .deserialize(FIRST_JSON))
+            .isEqualTo(FIRST);
+    }
+
+    @Test
+    void shouldThrowWhenDeserializeUnknownEvent() {
+        assertThatThrownBy(() -> JsonGenericSerializer.of()
+            .deserialize(FIRST_JSON))
+            .isInstanceOf(JsonGenericSerializer.UnknownTypeException.class);
+    }
+
+    @Test
+    void serializeShouldHandleAllKnownEvents() throws Exception {
+        JsonGenericSerializer<BaseType, DTO<BaseType>> serializer = JsonGenericSerializer.of(
+                TestModules.FIRST_TYPE,
+                TestModules.SECOND_TYPE);
+
+        assertThatJson(
+                serializer.serialize(SECOND))
+            .isEqualTo(SECOND_JSON);
+
+        assertThatJson(
```

Comment:
I thought a test should have just one assert?

Reply:
You should have only one cause of failure, here: its the capacity to to handle multiple serializers.

Reply:
it's not "one assert" but "one action" and "one goal".
I agree that we have two actions in this one and it's not ok.
@ParameterizedTest to the rescue?

Reply:
One action yes, sorry. @ParameterizedTest could make the trick then yes :)